File: WOT-iOS/WOT-iOS/ui/WOTTankListViewController/WOTTankConfigurationViewController/layout/WOTTankConfigurationFlowLayout.m
//
//  WOTTankConfigurationFlowLayout.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/30/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTTankConfigurationFlowLayout.h"

#define DEFAULT_SIZE CGSizeMake(44.0f,88.0f)

@implementation WOTTankConfigurationFlowLayout

- (void)prepareLayout {
    
    [super prepareLayout];
    
}

- (CGSize)itemSize {
    
    return DEFAULT_SIZE;
}

- (CGSize)collectionViewContentSize {

    NSInteger depth = 0;
    if (self.depthCallback) {
        
        depth = self.depthCallback();
    }
    
    NSInteger width = 0;
    if (self.widthCallback) {
        
        width = self.widthCallback();
    }
    
    CGSize result = CGSizeMake(width * self.itemSize.width,depth * self.itemSize.height);
    return result;
}


- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    CGSize defSize = DEFAULT_SIZE;
    UICollectionViewLayoutAttributes* attributes = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
    attributes.size = self.itemSize;
    attributes.center = CGPointMake(indexPath.row * defSize.width, indexPath.section * defSize.height);
    return attributes;
}

- (UICollectionViewLayoutAttributes *)initialLayoutAttributesForAppearingSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
    
    UICollectionViewLayoutAttributes *attributes = [self layoutAttributesForSupplementaryViewOfKind:kind atIndexPath:indexPath];
    return attributes;
}
- (UICollectionViewLayoutAttributes *)finalLayoutAttributesForDisappearingSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
 
    UICollectionViewLayoutAttributes *attributes = [self layoutAttributesForSupplementaryViewOfKind:kind atIndexPath:indexPath];
    return attributes;
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    
    NSMutableArray* elementsInRect = [NSMutableArray array];
    
    //iterate over all cells in this collection
    for(NSUInteger section = 0; section < [self.collectionView numberOfSections]; section++) {
        
        for(NSUInteger row = 0; row < [self.collectionView numberOfItemsInSection:section]; row++) {
            
            
            CGFloat x = row*self.itemSize.width;
            CGFloat y = section*self.itemSize.height;
            
            if (self.layoutPreviousSiblingNodeChildrenCountCallback) {
                
                //used with lines
                //NSInteger layoutSiblingChildrenCount = row;
                NSInteger layoutSiblingChildrenCount = self.layoutPreviousSiblingNodeChildrenCountCallback([NSIndexPath indexPathForRow:row inSection:section]);
                x = layoutSiblingChildrenCount * self.itemSize.width;
            }
            
            //this is the cell at row j in section i
            CGRect cellFrame = CGRectMake(x,
                                          y,
                                          self.itemSize.width,
                                          self.itemSize.height);
            
            //see if the collection view needs this cell
            if(CGRectIntersectsRect(cellFrame, rect)) {

                //create the attributes object
                NSIndexPath* indexPath = [NSIndexPath indexPathForRow:row inSection:section];
                UICollectionViewLayoutAttributes* attr = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
                
                //set the frame for this attributes object
                attr.frame = cellFrame;
                [elementsInRect addObject:attr];
            }
        }
    }
    
    return elementsInRect;
}

- (BOOL) shouldInvalidateLayoutForBoundsChange:(CGRect)newBound {
    
    return YES;
}



/*
 see: https://goo.gl/zNUcS1
 
    Has parent	Has child	Has next sibling	Has prev sibling    Binary  Dec
 1	0           1           0                   0                   0100	4
 2	1           1           1                   0                   1110	14
 3	0           1           1                   1                   0111	7
 4	0           1           0                   1                   0101	5
 5	1           1           0                   0                   1100	12
 6	1           0           0                   0                   1000	8
 7	0           0           1                   1                   0011	3
 8	0           0           0                   0                   0000	0

 */


@end

File: WOT-iOS/WOT-iOS/ui/WOTTankModuleTreeViewController/layout/WOTTankConfigurationFlowLayout.swift
//
//  WOTTankConfigurationFlowLayout.swift
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 7/26/18.
//  Copyright © 2018 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc
protocol WOTTankConfigurationFlowLayoutProtocol: NSObjectProtocol {
    var depthCallback: (() -> Int)? { get set }
    var widthCallback: (() -> Int)? { get set }
    var layoutPreviousSiblingNodeChildrenCountCallback: ((IndexPath) -> (Int))? { get set }
}

class WOTTankConfigurationFlowLayout: UICollectionViewFlowLayout, WOTTankConfigurationFlowLayoutProtocol {
    var depthCallback: (() -> Int)?
    var widthCallback: (() -> Int)?
    var layoutPreviousSiblingNodeChildrenCountCallback: ((IndexPath) -> (Int))?

    override open var itemSize: CGSize {
        get {
            return Constants.itemSize
        }
        set {

        }
    }

    override open var collectionViewContentSize: CGSize {
        get {
            let relativeSize = self.relativeSize()
            let itemSize = self.itemSize
            let depth = relativeSize.width * itemSize.width
            let width = relativeSize.height * itemSize.height
            return CGSize(width: width, height: depth)
        }
        set {

        }
    }

    override open func initialLayoutAttributesForAppearingSupplementaryElement(ofKind elementKind: String, at elementIndexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        return self.layoutAttributesForSupplementaryView(ofKind: elementKind, at: elementIndexPath)
    }

    override open func finalLayoutAttributesForDisappearingSupplementaryElement(ofKind elementKind: String, at elementIndexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        return self.layoutAttributesForSupplementaryView(ofKind: elementKind, at: elementIndexPath)
    }

    override open func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return true
    }

    override open func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        guard let collectionView = self.collectionView else {
            return nil
        }

        var result = [UICollectionViewLayoutAttributes]()
        for section in 0 ..< collectionView.numberOfSections {
            for row in 0 ..< collectionView.numberOfItems(inSection: section) {
                let indexPath = IndexPath(row: row, section: section)
                if let attrs = treeLayoutCellAttrubutes(indexPath: indexPath, rect: rect) {
                    result.append(attrs)
                }
            }
        }
        return result
    }
}

extension WOTTankConfigurationFlowLayout {

    fileprivate struct Constants {
        static let itemSize = CGSize(width: 44, height: 88)
    }

    fileprivate func relativeSize() -> CGSize {
        let width = CGFloat(self.width())
        let depth = CGFloat(self.depth())
        return CGSize(width: width, height: depth)
    }

    func depth() -> Int {
        guard let block = self.depthCallback else {
            return 0
        }

        return block()
    }

    func width() -> Int {
        guard let block = self.widthCallback else {
            return 0
        }
        return block()
    }

    private func siblingChildrenCount(indexPath: IndexPath) -> Int {
        guard let block = self.layoutPreviousSiblingNodeChildrenCountCallback else {
            return indexPath.row
        }
        return block(indexPath)
    }

    private func cellFrame(indexPath: IndexPath) -> CGRect {

        let itemSize = self.itemSize
        let siblingChildrenCount = self.siblingChildrenCount(indexPath: indexPath)
        let x = itemSize.width * CGFloat(siblingChildrenCount)
        let y = itemSize.height * CGFloat(indexPath.section)
        return CGRect(x: x, y: y, width: itemSize.width, height: itemSize.height)

    }

    private func treeLayoutCellAttrubutes(indexPath: IndexPath, rect: CGRect) -> UICollectionViewLayoutAttributes? {
        let cellFrame = self.cellFrame(indexPath: indexPath)
        if cellFrame.intersects(rect) {
            let attrs = UICollectionViewLayoutAttributes(forCellWith: indexPath)
            attrs.frame = cellFrame
            return attrs
        }
        return nil
    }
}
/*
 see: https://goo.gl/zNUcS1

 Has parent    Has child    Has next sibling    Has prev sibling    Binary  Dec
 1    0           1           0                   0                   0100    4
 2    1           1           1                   0                   1110    14
 3    0           1           1                   1                   0111    7
 4    0           1           0                   1                   0101    5
 5    1           1           0                   0                   1100    12
 6    1           0           0                   0                   1000    8
 7    0           0           1                   1                   0011    3
 8    0           0           0                   0                   0000    0

 */
